utils/compare_ds.py

```diff
@@ -0,0 +1,274 @@
+
+
+def find_benchmark(root, id_):
+    for component in root.findall("ds:component", ns):
+        benchmark = component.find("xccdf:Benchmark[@id='%s']" % (id_), ns)
+        if benchmark is not None:
+            return benchmark
+    return None
+
+
+def find_oval_definition(oval_doc, def_id):
+    definitions = oval_doc.find("oval:definitions", ns)
+    definition = definitions.find("oval:definition[@id='%s']" % (def_id), ns)
+    return definition
+
+
+def find_oval_test(oval_doc, test_id):
+    tests = oval_doc.find("oval:tests", ns)
+    test = tests.find("*[@id='%s']" % (test_id))
+    return test
+
+
+def definition_to_elements(definition):
+    criteria = definition.find("oval:criteria", ns)
+    elements = []
+    for child in criteria.iter():  # iter recurses
+        if child.tag == "{%s}criteria" % (ns["oval"]):
+            operator = child.get("operator")
+            elements.append(("criteria", operator))
+        elif child.tag == "{%s}criterion" % (ns["oval"]):
+            test_id = child.get("test_ref")
+            elements.append(("criterion", test_id))
+        elif child.tag == "{%s}extend_definition" % (ns["oval"]):
+            extend_def_id = child.get("definition_ref")
+            elements.append(("extend_definition", extend_def_id))
+    return elements
+
+
+def print_offending_elements(elements, sign):
+    for thing, atrribute in elements:
+        print("%s %s %s" % (sign, thing, atrribute))
+
+
+def compare_oval_definitions(
+        old_oval_def_doc, old_oval_def_id, new_oval_def_doc, new_oval_def_id):
+    old_def = find_oval_definition(old_oval_def_doc, old_oval_def_id)
+    new_def = find_oval_definition(new_oval_def_doc, new_oval_def_id)
+    old_els = definition_to_elements(old_def)
+    new_els = definition_to_elements(new_def)
+    for x in old_els.copy():
+        for y in new_els.copy():
+            if x[0] == y[0] and x[1] == y[1]:
+                old_els.remove(x)
+                new_els.remove(y)
+    if old_els or new_els:
+        print("OVAL definition %s differs:" % (old_oval_def_id))
+        print("--- old datastream")
+        print("+++ new datastream")
+        print_offending_elements(old_els, "-")
+        print_offending_elements(new_els, "+")
+
+
+def compare_ovals(
+        old_rule, new_rule, old_oval_defs, new_oval_defs, show_diffs):
+    old_oval_ref = old_rule.find(
+        "xccdf:check[@system='%s']" % (ssg.constants.oval_namespace), ns)
+    new_oval_ref = new_rule.find(
+        "xccdf:check[@system='%s']" % (ssg.constants.oval_namespace), ns)
+    rule_id = old_rule.get("id")
+    if (old_oval_ref is None and new_oval_ref is not None):
+        print("New datastream adds OVAL for rule '%s'." % (rule_id))
+    elif (old_oval_ref is not None and new_oval_ref is None):
+        print("New datastream is missing OVAL for rule '%s'." % (rule_id))
+    elif (old_oval_ref is not None and new_oval_ref is not None):
+        old_check_content_ref = old_oval_ref.find(
```

Comment:
Please extract this into a separate function with a good name, we have some serious `if` nesting here.